### ISSUE TITLE
Faiss use case in keep changing data

### ISSUE THREAD

**Issue by billyean** [facebookresearch/faiss#867]
# Summary

I understand that Faiss doesn't allow concurrent add and it's not recommended that using update mode on faiss. But we're on the POC that evaluate the use case of Faiss. we have a vectors that keep changing every 5 minutes, we want to build a high performance vector query service on it. Is there any way we can use Faiss on it; one way it our service keep restart and regenerate indices but this still add latency between restart. This lead to a couple of questions, what's the data structure of indices, can we persist indices somewhere and read them from file for saving the time to recreate all indices from the beginning.

# Platform

OS: 

Faiss version: 

Faiss compilation options: 

Running on:
- [ ] CPU
- [ ] GPU

Interface: 
- [ ] C++
- [ ] Python

# Reproduction instructions

**Comment by beauby** [comment]
The data structure of indexes depends on the index. 
You can indeed save indexes to file.

Answering your general question is difficult without knowing the specific constraints you are dealing with.

**Comment by billyean** [comment]
What we want to do is replace our current Solr index query with Faiss. Current solr has documents that will be changed periodically. We managed to move our vector to a mysql table, the change for the data will be changed to mysql table. We want to use Faiss to implement a high performance(lower latency than solr query and not lower throughput than solr) query. The problem for us is the data keeps changing(We have a data pipeline to write the changed data).  The service we're building need to restart when data changed, and if we need build index from scratch, it takes a couple of seconds, this is too much for us, we want to 
1. Lower the latency when service restart.
2. Ultimate goal is a service that can run concurrent with the data pipeline

**Comment by mdouze** [comment]
So you need an index type that can add / remove vectors. That could be an `IndexIVF` with a `direct_map`. 
Faiss does not support concurrent add / search. 
But you could use two indexes that you swap periodically: one in the foreground that handles all searches, one in the background that you perform adds on. If you swap 5 minutes, changes would be taken into account with that latency.

**Comment by LiberiFatali** [comment]
In this case, I have tested using a lock between add / remove vectors (for example: threading.RLock()). It works fine

**Comment by mdouze** [comment]
no activity. Closing.

**Comment by Talgin** [comment]
Hi,
We created an IndexIVF where we store vectors with ids. We used add_with_ids() method, and after inserting all the vectors (~20 million) we saved our index.
Now we have new vectors (~400 000). How can we add these vectors to an existing index? 
We tried to read previous index and use add_with_ids() on it, but some of the indices were changed after adding new vectors and ids. We need to keep ids and corresponding vectors of the previous index and add new ids and vectors.
